### DISCUSSION TITLE
Setting login timeout for mssql

### DISCUSSION THREAD

**Discussion by johanrex** [sqlalchemy/sqlalchemy#5874] [Usage Questions] [answered]
I sometimes get a login timeout from the mssql odbc driver when calling engine.connect(). From reading about this I have understood that there are two different timeouts, one for connecting and one for querying. This is about the login timeout. But I'm not sure how to set this from sqlachemy so it gets passed all the way down to the server. What there a preferred way of handling this? Any ideas you could think about would be helpful. 

sample code:
```
    with engine.connect() as conn:
        import pandas as pd

        df = pd.read_sql_query('select 8', conn)

```

Error message:
`(pyodbc.OperationalError) ('HYT00', '[HYT00] [Microsoft][ODBC Driver 17 for SQL Server]Login timeout expired (0) (SQLDriverConnect)')
`
When the error occurs the code is running in a docker container in Azure Kubernetes Services. Using python v3.8.
Azure SQL Server.

**Comment by zzzeek** [comment, marked as answer]
hiya -

you're using pyodbc so here is where you can find all the pyodbc connect parameters:

https://github.com/mkleehammer/pyodbc/wiki/The-pyodbc-Module

here's the docs for the "timeout" parameter:

https://github.com/mkleehammer/pyodbc/wiki/The-pyodbc-Module#timeout

to send parameters to the DBAPI from SQLAlchemy's engine we may use the [connect_args](https://docs.sqlalchemy.org/en/13/core/engines.html#special-keyword-arguments-passed-to-dbapi-connect) parameter, such as:

    e = create_engine("mssql+pyodbc://...", connect_args={"timeout": 30})

Most trivial "connect_args" may also be passed in the query string of the URL (that is, after the ?), for the pyodbc driver though I believe these get bundled into the ODBC connect string,  that is, it is going past pyodbc into the ODBC driver - from reading https://www.connectionstrings.com/all-sql-server-connection-string-keywords/ we'd want to use "Connect Timeout", so for this SQLAlchemy dialect we would use "Connect+Timeout", like this:

    e = create_engine("mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/test?driver=ODBC+Driver+13+for+SQL+Server&Connect+Timeout=30")


Below illustrates how to see the DSN string that this will generate so you can experiment yourself:

```python
>>> from sqlalchemy import create_engine
>>> e = create_engine("mssql+pyodbc://scott:tiger^5HHH@mssql2017:1433/test?driver=ODBC+Driver+13+for+SQL+Server&Connect+Timeout=30")
>>> print(e.dialect.create_connect_args(e.url))
[['DRIVER={ODBC Driver 13 for SQL Server};Server=mssql2017,1433;Database=test;UID=scott;PWD=tiger^5HHH;Connect Timeout=30'], {}]

```

**Reply by johanrex** [reply]
Wow, this is an impressively thorough answer. I'm very grateful. Thank you.

**Comment by zzzeek** [comment]
well the URL aspect of this is not really well documented for the pyodbc driver so we should improve the docs

**Reply by johanrex** [reply]
Yes, I agree wholeheartedly. From an end user perspective it's not obvious what settings are handled by the driver, pyodbc or sqlalchemy. And in the case that it's handled by the driver what the setting is even called and how to propagate it all the way through the whole stack. 

It may be obvious if you're already well versed in this architecture. But nobody is in the beginning. Understanding the whole setup takes some reading. 

Thanks again!

**Comment by zzzeek** [comment]
I improved the section at https://docs.sqlalchemy.org/en/13/core/engines.html#custom-dbapi-connect-arguments-on-connect-routines to introduce the topic a bit better as it was skipping over the "URL query string" part before.   Also it surprised me that the pyodbc driver is in fact taking all the query string elements and putting them into the DSN string, though I can see why it works that way.

**Reply by sshevlyagin** [reply]
This discussion is gold, adding the 2.0 link to make life easier https://docs.sqlalchemy.org/en/20/core/engines.html#custom-dbapi-args
